velox/functions/prestosql/window/Rank.cpp

```diff
@@ -41,25 +50,68 @@ class RankFunction : public exec::WindowFunction {
       vector_size_t resultOffset,
       const VectorPtr& result) override {
     int numRows = peerGroupStarts->size() / sizeof(vector_size_t);
-    auto* rawValues = result->asFlatVector<int64_t>()->mutableRawValues();
     auto* rawPeerStarts = peerGroupStarts->as<vector_size_t>();
+
+    void* rawValues = NULL;
+    switch (R) {
+      case RankType::kRank:
+      case RankType::kDenseRank: {
+        auto* resultRawValues =
+            result->asFlatVector<int64_t>()->mutableRawValues();
+        rawValues = resultRawValues;
+        break;
+      }
+      case RankType::kPercentRank: {
+        auto* resultRawValues =
+            result->asFlatVector<double>()->mutableRawValues();
+        rawValues = resultRawValues;
+        break;
+      }
+    }
+
     for (int i = 0; i < numRows; i++) {
       auto start = rawPeerStarts[i];
-      if (start != currentPeerGroupStart_) {
-        currentPeerGroupStart_ = start;
-        rank_ += previousPeerCount_;
-        previousPeerCount_ = 0;
+      switch (R) {
+        case RankType::kRank: {
+          if (start != currentPeerGroupStart_) {
+            currentPeerGroupStart_ = start;
+            rank_ += previousPeerCount_;
+            previousPeerCount_ = 0;
+          }
+          ((int64_t*)rawValues)[resultOffset + i] = rank_;
+          previousPeerCount_ += 1;
+          break;
+        }
+        case RankType::kDenseRank: {
+          if (start != currentPeerGroupStart_) {
+            currentPeerGroupStart_ = start;
+            rank_ += previousPeerCount_;
+            previousPeerCount_ = 0;
+          }
+          ((int64_t*)rawValues)[resultOffset + i] = rank_;
+          previousPeerCount_ = 1;
+          break;
+        }
+        case RankType::kPercentRank: {
+          if (start != currentPeerGroupStart_) {
+            currentPeerGroupStart_ = start;
+            rank_ += previousPeerCount_;
+            previousPeerCount_ = 0;
+          }
+          ((double*)rawValues)[resultOffset + i] =
+              double(rank_ - 1) / (numPartitionRows_ - 1);
+          previousPeerCount_ += 1;
```

Comment:
Even previousPeerCount_ += 1 is common across all cases. I feel like that should move after the switch statement.

The switch statement should just have a single line for setting the value in rawValues.

Reply:
Thanks for the suggestion but previousPeerCount_ is used as a flag in kDenseRank and is set to 1. So moving this outside the switch would require having the updation of rank_ variable in the switch case, with something like ``rank += (previousPeerCount_ != 0)``. Retained previousPeerCount updation in switch since that seems more concise.

Reply:
There might be another way to code this logic:

We always count previousPeerCount_ += 1 in the code here.

if (start != currentPeerGroupStart_) {
        currentPeerGroupStart_ = start;
*****************************************************************
Change the code here to only increment rank_ by 1 if kDenseRank
        rank_ += previousPeerCount_;
        previousPeerCount_ = 0;
      }

That would make the logic a bit simpler also.

Reply:
Thanks for the suggestion! For kDenseRank though, rank_ is updated by 1 only when previousPeerCount_ is not 0 so changed accordingly.